src/main/java/io/appium/java_client/AppiumFluentWait.java

```diff
@@ -222,32 +247,52 @@ public <V> V until(Function<? super T, V> isTrue) {
             // Check the timeout after evaluating the function to ensure conditions
             // with a zero timeout can succeed.
             if (end.isBefore(getClock().instant())) {
-                String message = getMessageSupplier() != null ? getMessageSupplier().get() : null;
-
-                String timeoutMessage = String.format(
-                        "Expected condition failed: %s (tried for %d second(s) with %s interval)",
-                        message == null ? "waiting for " + isTrue : message,
-                        getTimeout().getSeconds(), getInterval());
-                throw timeoutException(timeoutMessage, lastException);
+                handleTimeoutException(lastException, isTrue);
             }
 
-            try {
-                Duration interval = getInterval();
-                if (pollingStrategy != null) {
-                    final IterationInfo info = new IterationInfo(iterationNumber,
-                            Duration.between(start, getClock().instant()), getTimeout(),
-                            interval);
-                    interval = pollingStrategy.apply(info);
-                }
-                getSleeper().sleep(interval);
-            } catch (InterruptedException e) {
-                Thread.currentThread().interrupt();
-                throw new WebDriverException(e);
-            }
+            Duration interval = getIntervalWithPollingStrategy(start, iterationNumber);
+            sleepUninterruptible(interval);
+
             ++iterationNumber;
         }
     }
 
+    private <V> void handleTimeoutException(Throwable lastException, Function<? super T, V> isTrue) {
+        String message = Optional.ofNullable(getMessageSupplier()).map(Supplier::get).orElseGet(() -> "waiting for " + isTrue);
+
+        String timeoutMessage = String.format(
+                "Expected condition failed: %s (tried for %s millis with an interval of %s millis)",
+                message,
+                getTimeout().toMillis(),
+                getInterval().toMillis()
+        );
+
+        throw timeoutException(timeoutMessage, lastException);
+    }
+
+    private Duration getIntervalWithPollingStrategy(Instant start, long iterationNumber) {
+        Duration interval = getInterval();
+        return Optional.ofNullable(pollingStrategy)
+                .map(strategy -> {
+                    final IterationInfo info = new IterationInfo(
```

Comment:
there is no need for the intermediate `info` variable. Also consider using `var` instead of explicit typedefs - this reduces the amount of the boilerplate code

Reply:
Why? After we apply the info to the strategy. Do you mean directly constructing the `info` variable into the `strategy#apply`?

Reply:
> Do you mean directly constructing the info variable into the strategy#apply

yes

Reply:
Done